[STT-1301] Refactor display field config types and field rendering

eos87 opened the pull request:
### Purpose
This refactor makes the field rendering logic safer and easier to maintain. The old implementation relied heavily on `any` types and ad-hoc checks, which made it harder to understand and prone to subtle bugs. By introducing a proper `IDisplayFieldsConfig` union and linking field names directly to `IArticle`, the code is now easier to follow and less error-prone. 
**No UI changes are expected.**

> [!NOTE]  
> Added some documentation here https://sofab.atlassian.net/wiki/x/AwAd0QI

### What has changed
- Reworked `IDisplayFieldsConfig` into a single, clear union supporting simple fields (`keyof IArticle`), separators (`'/' | '//' | '-'`), arrays, styled wrappers, and component overrides.
- Added types: `Separator`, `ComponentName`, `StyledField`, `ComponentField`.
- Added `charcount?: number` to `IArticle`.
- Typed default meta arrays in `WireListItem` (`DEFAULT_META_FIELDS`, `DEFAULT_COMPACT_META_FIELDS`) as `IDisplayFieldsConfig`.
- Converted `VersionType` to TypeScript props (`{ value: string }`) and removed `prop-types`.
- Refactored `fields/index.tsx` for readability and type safety:
  - small, named type guards: `isSeparator`, `isStringField`, `isStyledField`, `isComponentField`
  - explicit `FieldResult` / `FieldComponentsProps`
  - cleaner separator handling 

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] This pull request is not adding new forms that use redux
- [x] This pull request is adding missing TypeScript types to modified code segments where it's easy to do so with confidence
- [ ] This pull request is replacing `lodash.get` with optional chaining for modified code segments

Belongs to [STT-1301]


[STT-1301]: https://sofab.atlassian.net/browse/STT-1301?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

eos87 committed:
Refactor display field config types and field rendering

Refactors the IDisplayFieldsConfig type to use explicit union types for separators, styled fields, and component overrides. Updates field rendering logic in FieldComponents to use new type guards and interfaces, improving type safety and clarity. Removes PropTypes from VersionType and replaces with TypeScript interface.

STT-1301

eos87 committed:
Refactor field config types and add charcount to IArticle

Updated IDisplayFieldsConfig to support nested arrays and enforce keyof IArticle for field references. Added 'charcount' to IArticle interface. Updated related components to use stricter types and improved type safety for field configs.

STT-1301

eos87 commented:
A second PR will be added to allow a new component type like what we have for version type in order to resolve STT-1301

eos87 committed:
Simplified the type assertion

thecalcc commented on assets/interfaces/configs.ts:
```diff
@@ -1,23 +1,31 @@
+export type ComponentName = 'version_type' | string;
+
+// component override: { field: "version", component: "version_type" }
+export interface ComponentField {
```
you can also do a generic argument here with default `never` that allows you to have a proper type for styles if you want to have it more specific. Also why not just put it as `CSSProperties`?

thecalcc commented on assets/wire/components/fields/index.tsx:
```diff
@@ -9,11 +9,37 @@ import {Embargo} from './Embargo';
+const isSeparator = (f: FieldConfig): f is Separator =>
+    typeof f === 'string' && (ALLOWED_SEPARATORS).includes(f as Separator);
+
+const isStringField = (f: FieldConfig): f is keyof IArticle => typeof f === 'string' && !isSeparator(f);
```
maybe we can have a comment here to explain why you check `!isSeperator(f)` since separator is also a string. or another function name like `isGenericString`

eos87 replied:
not needed - I think. The idea is that when it is a separator we treat it differently as we render it in a span tags and stop it there.

thecalcc commented on assets/wire/components/fields/index.tsx:
```diff
@@ -61,46 +89,10 @@ export function FieldComponents({config, item, fieldProps = {}}: any) {
-
-        return null;
-    } else if (Array.isArray(field) && field.length > 0) {
+function getComponentForField(item: IArticle, fieldConfig: FieldConfig): any {
```
could use a comment that it's recursive

thecalcc commented on assets/interfaces/configs.ts:
```diff
@@ -1,23 +1,31 @@
+export interface StyledField {
+    field: IDisplayFieldsConfig;
+    styles: { [key: string]: string | number };
+    component?: never;
```
why is `component` `never`?

eos87 replied:
leftover. fixed

thecalcc commented on assets/wire/components/fields/index.tsx:
```diff
@@ -9,11 +9,37 @@ import {Embargo} from './Embargo';
+}
+
+interface FieldComponentsProps {
+    config: FieldConfig | FieldConfig[];
```
i think we don't use `type[]` rather `Array<type>`

thecalcc commented on assets/wire/components/fields/index.tsx:
```diff
@@ -9,11 +9,37 @@ import {Embargo} from './Embargo';
-const ALLOWED_SEPARATORS = ['/', '//', '-'];
+interface FieldResult {
+    key: string;
+    Component: React.ComponentType<any>;
```
could probably be `React.ReactNode` to avoid `any`

eos87 replied:
no because `React.ReactNode` is a rendered value and what we need is the component itself before being rendered

thecalcc commented on assets/wire/components/fields/index.tsx:
```diff
@@ -9,11 +9,37 @@ import {Embargo} from './Embargo';
+const isStringField = (f: FieldConfig): f is keyof IArticle => typeof f === 'string' && !isSeparator(f);
+
+const isObjectCfg = (f: FieldConfig): f is StyledField | ComponentField =>
+    typeof f === 'object' && f !== null && !Array.isArray(f);
```
could you make this one cleaner like `typeof f === 'object' && !Array.isArray(f ?? [])` and since you're doing a strict check for not null probably f should be `FieldConfig | null`

thecalcc commented on assets/interfaces/configs.ts:
```diff
@@ -1,23 +1,31 @@
+    component?: never;
+}
+
+export type IDisplayFieldsConfig =
```
i guess it's not possible to enhance this type to add a unifying property like `type` or `kind` which would make type narrowing easier later, since this is coming from the backend?

eos87 replied:
sorry I'm not following here. Can you please elaborate a little bit more how those properties would be used?

thecalcc reviewed (commented):
looks good, but I guess my main question is readability - I never liked those functions that do type narrowing based on some properties, if we could add this `type`/`kind` property to `IDisplayFieldsConfig` it would improve the code a lot I think, but if it can't be added easily i guess readability of these functions would be the next best option

eos87 committed:
Refactor field rendering types and FieldComponents logic

Introduces FieldRenderProps and FieldRenderComponent types for improved type safety in field rendering

STT-1301

eos87 committed:
Update isObjectCfg to accept null as argument

The isObjectCfg function now accepts FieldConfig or null, improving type safety and preventing potential runtime errors when null values are passed.

STT-1301

eos87 commented:
> looks good, but I guess my main question is readability - I never liked those functions that do type narrowing based on some properties, if we could add this `type`/`kind` property to `IDisplayFieldsConfig` it would improve the code a lot I think, but if it can't be added easily i guess readability of these functions would be the next best option

Hi @thecalcc. I get your concern, but in this case the small, named type guards improve readability versus the previous chain of if/else checks. A few reasons: 
- The intent is explicit (like reading plain English) 
- Single responsibility
- They would make the code easier to test (in the near future or next PR ;))

Thoughts?

tomaskikutis commented:
[Discriminated unions](https://www.typescriptlang.org/play/?#code/PTAEBEEsGcGMCdIFtIDsCGAXApgE1AK6qQD2qoAZifKNABboAO2omAns9ALABQ7zoAMKR4sADYsAvKADeoANZpcALlAAiWCPHY1AblDx0uSAWirUBJACNsNAL67e-FgGUAjgXTwpshUtVq0B5eOvrQkLjY5pY29o58HK4MAtLCohKgAD6g7p7e8bwgoAAqiQC00OgULBREsJik5KZoAOasdCzGcIgoGKiYvLWo9Y2gLdiYAILe6AAU9ExROcnYAJTR1rayvKC0AO6QmLB0oPMrAHSKqLir2zy7u7Do0CwaWhJqyjsPj2TQJBJzmISC0zotzuFIrcigAFeAkZjwdigADkkOwKNAuBI2GgoFQJEwoGwAA8YESyKxEqi0toUedZgAmADMzIAnKtvj9vJgCPByABZLB0c4wgCSoAAVLQLoZjKYpdLGfEfk8XuognkdF97j9QLA-gDsECQWDmOc5SZoNCwHCEbZkSjLaZMdjcfjCcSydAKeRnKjciF6UzWRyuQ8eXzyAtzejFaBld87LxkzxCmAAKqVca8A2of6A4Gg8ZTGazORXFTqTTpHQAGgMRitqgArKA7KtVvpQEUAH4AdgAHOcWwAWXOGwumkvTbBzCv+DXBbxqBvo1Sj9ud7u7IoARgAbEA) are always a better choice when possible. When not possible or would be especially time consuming to use them, then casting to a type based on some properties is best that remains.

eos87 committed:
Merge remote-tracking branch 'origin/develop' into hg/STT-1301-improve-fields-config-rendering
